### PR TITLE
FEATURE: TNT-1284 Dashboard permissions scenarios

### PR DESCRIPTION
**Pull request by tomassatinsky**
---

Supported PR commands:

| Command                                               | Description                                                |
| ----------------------------------------------------- | ---------------------------------------------------------- |
| `ok to test`                                          | Re-run standard checks                                     |
| `extended check sonar`                                | SonarQube tests                                            |
| `extended check plugins`                              | Dashboard plugins tests                                    |
| `extended test - backstop`                            | BackstopJS tests                                           |
| **E2E Cypress tests commands - TIGER**                |                                                            |
| `extended test - tiger-cypress - isolated <testName>` | Run isolated tests running against recorded Tiger backend. |
| `extended test - tiger-cypress - record <testName>`   | Create a new recording for isolated Tiger tests.           |
| **E2E Cypress tests commands - BEAR**                 |                                                            |
| `extended test - cypress - isolated <testName>`       | Run isolated tests running against recorded Bear backend.  |
| `extended test - cypress - record <testName>`         | Create a new recording for isolated Bear tests.            |

`<testName>` in cypress commands is used to filter specfiles. Example, to run record with BEAR backend

-   Against `dashboard.spec.ts` and `drilling.spec.ts`, execute command `extended test - cypress - record dashboard,drilling`
-   Against all specfiles, execute command `extended test - cypress - record` or `extended test - cypress - record *`

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `extended test - backstop` passes
-   [ ] `extended test - tiger-cypress - record` to record new mapping files (Tiger BE)
-   [ ] `extended test - cypress - record` to record new mapping files (Bear BE)
-   [ ] `extended test - tiger-cypress - isolated` passes
-   [ ] `extended test - cypress - isolated` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11286/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13374/ : FAILURE in 10m 12s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11301/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13383/ : FAILURE in 3m 47s

**Comment by scavnickyj** [comment]
extended test - tiger-cypress - isolated

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher/105/ : FAILURE in 4m 18s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11305/ : SUCCESS in 8s

File: libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts [hunk 1]
```diff
@@ -52,7 +52,9 @@ function* createDashboardSaveSharingContext(cmd: ChangeSharing): SagaIterator<Da
     };
     const dashboardToSave: IDashboardDefinition = {
         ...dashboardFromState,
-        ...newSharingProperties,
+        shareStatus: newSharingProperties.shareStatus,
+        isLocked: newSharingProperties.isLocked,
+        isUnderStrictControl: newSharingProperties.isUnderStrictControl,
```
Comment (scavnickyj) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts:57]: I can see that this omits `grantees..`. Is this intentional?

File: libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularGranteeUserItem.tsx [hunk 2]
```diff
@@ -32,24 +34,33 @@ export const GranularGranteeUserItem: React.FC<IGranularGranteeUserItemProps> =
...
+    const label = useMemo(() => {
         return getGranteeLabel(grantee, intl);
     }, [grantee, intl]);
 
-    const renderSubtitle = useMemo(() => email && email !== userName, [email, userName]);
+    const renderSubtitle = useMemo(() => email && email !== label, [email, label]);
+
+    const granteePossibilities = useMemo(
+        () => getGranteePossibilities(grantee, currentUserPermissions, isDashboardLocked),
+        [currentUserPermissions, isDashboardLocked],
```
Comment (scavnickyj) [inline libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularGranteeUserItem.tsx:45]: `grantee` dependency is missing

File: libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularPermissionItem.tsx [hunk 3]
```diff
@@ -32,8 +31,10 @@ const GranularPermissionSelectItem: React.FC<IGranularPermissionItemProps> = ({
...
         (permission: IGranularPermissionTypeItem) => {
             toggleDropdown();
-            handleSetSelectedPermission(permission.id);
-            onChange({ ...grantee, permissions: [permission.id] });
+            if (permission.id !== selectedPermission) {
+                handleSetSelectedPermission(permission.id);
+                onChange({ ...grantee, permissions: [permission.id] });
+            }
         },
         [grantee, toggleDropdown, handleSetSelectedPermission, onChange],
```
Comment (scavnickyj) [inline libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularPermissionItem.tsx:39]: `selectedPermission` is missing in dependencies

**Comment by scavnickyj** [comment]
extended test - cypress - isolated

**Comment by scavnickyj** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher/107/ : FAILURE in 6m 50s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11307/ : SUCCESS in 15s

File: libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts (hunk 1 above)
Comment (tomassatinsky) [reply libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts:57]: Definitely. This code is building dashboard to save so grantees should not be there. It was failing before the change as it tried to save the dashboard (which was in parent workspace) to the child workspace. For some reason, when the original and new dashboards are the same, it skips the saving (updateDashboard)

File: libs/sdk-ui-dashboard/src/presentation/topBar/buttonBar/button/selectors.ts [hunk 4]
```diff
@@ -16,40 +15,23 @@ import {
...
-    selectIsEditModeEnable,
+    selectDashboardEditModeDevRollout,
     selectCanEditDashboardPermission,
+    selectCanEditLockedDashboardPermission,
     selectDashboardLockStatus,
-    selectCanEditLockedDashboard,
-    (isEditModeEnable, canManageAnalyticalDashboard, isLocked, canEditLockedDashboard) =>
-        isEditModeEnable && canManageAnalyticalDashboard && (!isLocked || canEditLockedDashboard),
+    (isEditModeEnabled, canEditDashboardPermission, canEditLockedDashboardPermission, isLocked) =>
+        isEditModeEnabled && canEditDashboardPermission && (!isLocked || canEditLockedDashboardPermission),
```
Comment (tomassatinsky) [inline libs/sdk-ui-dashboard/src/presentation/topBar/buttonBar/button/selectors.ts:34]: FYI: I've removed the capability check as now both backends support permissions. And `canManageWorkspace` was replaced with canEditLockedDashboardPermission (which is `false` on Tiger and `canManageWorkspace` on Bear)

File: libs/sdk-ui-dashboard/src/presentation/topBar/buttonBar/button/selectors.ts [hunk 5]
```diff
@@ -104,25 +75,28 @@ export const selectIsCurrentDashboardVisibleInList = createSelector(
...
         isLocked,
+        isCurrentDashboardVisibleInList,
         isReadOnly,
         isInEditMode,
         isShareButtonHidden,
     ) =>
-        hasPermission &&
+        dashboardPermissionsEnabled &&
+        canShareDashboardPermission &&
+        (!isLocked || canShareLockedDashboardPermission) &&
```
Comment (tomassatinsky) [inline libs/sdk-ui-dashboard/src/presentation/topBar/buttonBar/button/selectors.ts:98]: In this function, again `canManageWorkspace` was replaced with canShareLockedDashboardPermission (which is based on SHARE dashboard permission on Tiger and `canManageWorkspace` WS permission on Bear)

File: libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularGranteeUserItem.tsx (hunk 2 above)
Comment (tomassatinsky) [reply libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularGranteeUserItem.tsx]: fixed

File: libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularPermissionItem.tsx (hunk 3 above)
Comment (tomassatinsky) [reply libs/sdk-ui-kit/src/Dialog/ShareDialog/ShareDialogBase/GranularPermissions/GranularPermissionItem.tsx]: fixed

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11311/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13390/ : FAILURE in 3m 23s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11312/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13392/ : FAILURE in 19m 36s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11319/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13396/ : FAILURE in 26m 54s

**Comment by tomassatinsky** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11328/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13403/ : FAILURE in 21m 33s

**Comment by tomassatinsky** [comment]
extended test - backstop

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11332/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13405/ : SUCCESS in 20m 52s

**Comment by yenkins** [comment]
Link to Storybook test result: https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-backstop-tests/1950/artifact/libs/sdk-ui-tests/backstop/output/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher/120/ : SUCCESS in 31m 31s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11333/ : SUCCESS in 7s

**Comment by tomassatinsky** [comment]
extended test - cypress - isolated

**Comment by yenkins** [comment]
# ✨ Cypress test results ✨

## 🎉 Great news! Tests were successful.

|Status|Name|Tests|Failures|Time|
|:--:|:--|:--:|:--:|--:|
|✅|**drilling.spec.ts**|7(2 skipped)|-|46.6220s|
|✅|**barChartDrilling.spec.ts**|4|-|7.2040s|
|✅|**attributeFilterButton.spec.ts**|9|-|32.7580s|
|✅|**dashboardDnD.spec.ts**|7|-|65.0850s|
|✅|**pivotTableRemoveOneFromMutipleAggregation.spec.ts**|1|-|7.7910s|
|✅|**pivotTableSizing.spec.ts**|1|-|5.3860s|
|✅|**pivotTableAggregationsMenuAndAddTotal.spec.ts**|5|-|21.8640s|
|✅|**pivotTableSizingReset.spec.ts**|15|-|43.1350s|
|✅|**dateFilter.spec.ts**|9|-|17.1080s|
|✅|**dashboard.spec.ts**|9|-|63.5840s|
|✅|**dashboardDependentFilter.spec.ts**|4|-|53.3760s|
|✅|**pivotTableRemoveAllFromMutipleAggregation.spec.ts**|1|-|7.9060s|
|✅|**headlineChartDrilling.spec.ts**|2|-|4.3740s|
|✅|**attributeFilterButtonWithSelection.spec.ts**|2|-|8.1800s|
|✅|**dashboardDateRelated.spec.ts**|8|-|74.6730s|
|✅|**parentChildFiltersWithPlaceholders.spec.ts**|3|-|18.8360s|
|✅|**attributeFilterButtonParentChild.spec.ts**|3|-|15.4880s|
|✅|**pivotTableRemoveOneAggregation.spec.ts**|1|-|7.4290s|
|✅|**dashboardFiltering.spec.ts**|18(2 skipped)|-|166.0600s|
|✅|**dashboardKPI.spec.ts**|8|-|158.7870s|

You can find full report [here](https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-docker/387/).

**Comment by yenkins** [comment]
# ✨ Cypress test results ✨

## 🎉 Great news! Tests were successful.

|Status|Name|Tests|Failures|Time|
|:--:|:--|:--:|:--:|--:|
|✅|**drilling.spec.ts**|7(2 skipped)|-|46.6220s|
|✅|**barChartDrilling.spec.ts**|4|-|7.2040s|
|✅|**attributeFilterButton.spec.ts**|9|-|32.7580s|
|✅|**dashboardDnD.spec.ts**|7|-|65.0850s|
|✅|**pivotTableRemoveOneFromMutipleAggregation.spec.ts**|1|-|7.7910s|
|✅|**pivotTableSizing.spec.ts**|1|-|5.3860s|
|✅|**pivotTableAggregationsMenuAndAddTotal.spec.ts**|5|-|21.8640s|
|✅|**pivotTableSizingReset.spec.ts**|15|-|43.1350s|
|✅|**dateFilter.spec.ts**|9|-|17.1080s|
|✅|**dashboard.spec.ts**|9|-|63.5840s|
|✅|**dashboardDependentFilter.spec.ts**|4|-|53.3760s|
|✅|**pivotTableRemoveAllFromMutipleAggregation.spec.ts**|1|-|7.9060s|
|✅|**headlineChartDrilling.spec.ts**|2|-|4.3740s|
|✅|**attributeFilterButtonWithSelection.spec.ts**|2|-|8.1800s|
|✅|**dashboardDateRelated.spec.ts**|8|-|74.6730s|
|✅|**parentChildFiltersWithPlaceholders.spec.ts**|3|-|18.8360s|
|✅|**attributeFilterButtonParentChild.spec.ts**|3|-|15.4880s|
|✅|**pivotTableRemoveOneAggregation.spec.ts**|1|-|7.4290s|
|✅|**dashboardFiltering.spec.ts**|18(2 skipped)|-|166.0600s|
|✅|**dashboardKPI.spec.ts**|8|-|158.7870s|

You can find full report [here](https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-cypress-bear-isolated-zuul-docker/387/).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-pull-request-dispatcher/121/ : SUCCESS in 26m 30s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11334/ : SUCCESS in 6s

File: libs/sdk-backend-bear/src/backend/workspace/dashboards/index.ts [hunk 6]
```diff
@@ -1035,15 +1035,23 @@ export class BearWorkspaceDashboards implements IWorkspaceDashboardsService {
...
             return {
-                canEditDashboard: workspacePermissions.canManageAnalyticalDashboard,
-                canShareDashboard: workspacePermissions.canManageACL,
+                canEditDashboard,
+                canEditLockedDashboard: canManageLockedDashboard,
+                canShareDashboard,
+                canShareLockedDashboard: canManageLockedDashboard,
                 canViewDashboard: true,
             };
         } catch (_e) {
```
Comment (ivanmjartan) [inline libs/sdk-backend-bear/src/backend/workspace/dashboards/index.ts:1049]: Remove unnecessary _e parameter

File: libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts [hunk 7]
```diff
@@ -52,7 +52,9 @@ function* createDashboardSaveSharingContext(cmd: ChangeSharing): SagaIterator<Da
     };
     const dashboardToSave: IDashboardDefinition = {
         ...dashboardFromState,
```
Comment (ivanmjartan) [inline libs/sdk-ui-dashboard/src/model/commandHandlers/dashboard/changeSharingHandler.ts:54]: Could you try Add handler tests

**Review by ivanmjartan — APPROVED** [review]

**Review by ivanmjartan — APPROVED** [review]

**Comment by ivanmjartan** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/11345/ : SUCCESS in 11s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13413/ : SUCCESS in 28m 57s

**Review by scavnickyj — APPROVED** [review]

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3428

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/297/ : FAILURE in 2m 35s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3564/ : SUCCESS in 32s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2412/ : SUCCESS in 20m 01s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13414/ : SUCCESS in 21m 46s

**Pull request merged**